src/Pos/Delegation/Cede/Logic.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Different logic depending on "MonadCede".
+
+module Pos.Delegation.Cede.Logic
+       (
+         getPskChain
+       , getPskChainInternal
+       , detectCycleOnAddition
+       , dlgReachesIssuance
+       ) where
+
+import           Universum
+
+import qualified Data.HashMap.Strict       as HM
+import qualified Data.HashSet              as HS
+
+import           Control.Lens              (uses, (%=))
+import           Pos.Crypto                (ProxySecretKey (..), PublicKey)
+import           Pos.DB                    (DBError (DBMalformed))
+import           Pos.Delegation.Cede.Class (MonadCedeRead (..))
+import           Pos.Delegation.Helpers    (isRevokePsk)
+import           Pos.Delegation.Types      (DlgMemPool)
+import           Pos.Types                 (ProxySKHeavy, StakeholderId, addressHash)
```

Comment:
It's a bad idea to import `Pos.Types` from delegation, because we have plans to move it deeper.